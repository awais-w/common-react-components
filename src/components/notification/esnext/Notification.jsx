import React from 'react';
import PropTypes from 'prop-types';
import bem from '@argos/utils/esnext/bem';
import Icon from '@cmc/icon/esnext/Icon';
import './Notification.scss';

const cn = bem({ prefix: 'cmc-', block: 'notification' });

const modes = ['info', 'warning', 'error'];
const alignments = ['top', 'middle'];

const propTypes = {
  iconAlignment: PropTypes.oneOf(alignments),
  mode: PropTypes.oneOf(modes),
};

const defaultProps = {
  mode: 'info',
  iconAlignment: 'middle',
};

const Notification = (props) => {
  const { children, mode, iconAlignment } = props;

  return (
    <div className={cn(null, mode, `icon-alignment-${iconAlignment}`)}>
      <Icon icon={mode} />
      <span className={cn('message')}>{children}</span>
    </div>
  );
};

Notification.propTypes = propTypes;
Notification.defaultProps = defaultProps;
export default Notification;
