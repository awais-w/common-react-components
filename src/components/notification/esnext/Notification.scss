@import '@cmc/common-styles/src/scss/_variables.scss';

.cmc-notification {
  width: 100%;
  padding: $base;
  display: flex;
  align-items: center;
  border-radius: $border-radius-base;
  @include font-standard;

  &--info {
    border: 1px solid $primary-color;
    background-color: $color-alert-blue;
  }

  &--warning {
    border: 1px solid $color-sainsburys;
    background-color: $color-alert-orange;
  }

  &--error {
    border: 1px solid $color-rust;
    background-color: $color-alert-red;
  }

  &__message {
    margin-left: $base;
    font-size: $font-size-base;
    font-weight: bold;
    width: 100%;
  }

  span ~ button {
    display: block;
    margin: $base auto 0 auto;
  }

  &.icon-alignment-top {
    i {
      place-self: start;
    }
  }

  &.icon-alignment-middle {
    i {
      place-self: center;
    }
  }
}
