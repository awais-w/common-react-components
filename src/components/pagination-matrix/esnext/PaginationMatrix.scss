@import '@cmc/common-styles/src/scss/_variables.scss';
@import './Matrix.scss';

.cmc-pagination-matrix {
  @include font-standard;

  &__matrix-grid {
    background-color: $color-grey-light;
    min-height: 100px;
    padding: $double;
    .cmc-notification--error {
      margin-bottom: $base;
    }
  }

  &__table {
    overflow: auto;
  }

  &__matrix-grid--highlight {
    @extend .cmc-pagination-matrix__matrix-grid;
    background-color: lighten($color-rust, 50%);
  }

  &__arrow-container {
    display: flex;
    justify-content: flex-end;
    margin-top: $base;
  }

  &__prev-arrow {
    button {
      margin-right: $base;
    }
  }

  &--disabled {
    @include disabled;
  }
}
