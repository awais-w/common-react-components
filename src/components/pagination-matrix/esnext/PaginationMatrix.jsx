import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import isEmpty from '@cmc/utils/esnext/isEmpty';
import Loading from '@cmc/loading/esnext/Loading';
import bem from '@argos/utils/esnext/bem';
import Button from '@cmc/cha-button/esnext/Button';
import Notification from '@cmc/cha-notification/esnext/Notification';
import './PaginationMatrix.scss';
import MatrixRow from './components/matrix-row/MatrixRow';
import MatrixHeadings from './components/matrix-headings/MatrixHeadings';

export function checkMatrixData(props, propName, componentName) {
  // get lengths
  const headingsLength = props.matrixData.headings && props.matrixData.headings.length;
  const sideHeadingsLength = props.matrixData.sideHeadings && props.matrixData.sideHeadings.length;
  const rowsLength = props.matrixData.rows && props.matrixData.rows.length;
  // check if data is empty or not
  const headingsEmpty = headingsLength === 0;
  const sideHeadingsEmpty = sideHeadingsLength === 0;
  const rowsEmpty = rowsLength === 0;
  // TODO: remove this variable and rather than assingning values to error we should return
  let error = null;
  // display error if no data is provided at all
  if (headingsEmpty && sideHeadingsEmpty && rowsEmpty) {
    error = new Error(
      `Invalid prop \`${propName}\` supplied to` +
        ` \`${componentName}\`. Validation failed. No data provided to ${propName}`,
    );
  }

  // display error if both headings and rows are empty
  if (headingsEmpty && rowsEmpty) {
    error = new Error(
      `Invalid prop \`${propName}\` supplied to` +
        ` \`${componentName}\`. Validation failed. No headings or rows provided to ${propName}`,
    );
  }

  // display error if there are side headings but not enough rows to match
  if (!sideHeadingsEmpty && props.matrixData.sideHeadings !== undefined) {
    if (sideHeadingsLength !== rowsLength) {
      error = new Error(
        `Invalid prop \`${propName}\` supplied to` +
          ` \`${componentName}\`. Validation failed. Sideheadings and rows should match up in ${propName}`,
      );
    }
  }

  if (!rowsEmpty && !headingsEmpty) {
    props.matrixData.rows.forEach((row) => {
      if (row.length !== headingsLength) {
        error = new Error(
          `Invalid prop \`${propName}\` supplied to` +
            ` \`${componentName}\`. Validation failed. Headings and rows should match up in ${propName}`,
        );
      }
    });
  }

  return error || null;
}

const propTypes = {
  /**
   * This is the data provided to the matrix
   * TODO: Consider validating deeper than this
   */
  matrixData: checkMatrixData,
  /**
   * Maximum number of cells which should be displayed before pagination
   */
  visibleColumns: PropTypes.number,
  /**
   * Show 'previous/next' buttons even if they need to be greyed out
   */
  alwaysShowPaginationButtons: PropTypes.bool,
  /**
   * Click handler for cells
   */
  onClickCell: PropTypes.func,
  /**
   * Determines whether the component is waiting or not
   */
  isLoading: PropTypes.bool,
  /**
   * Will take the place of the increment function if set
   */
  onNextBatchRequest: PropTypes.func,
  /**
   * The text displayed next to the pagination arrows
   */
  nextPageText: PropTypes.string,
  /**
   * Should we show the next button
   */
  shouldShowNextButton: PropTypes.bool,
  /**
   * An error message to display
   */
  error: PropTypes.string,
  /**
   * The column offset of the matrix to display
   */
  columnOffset: PropTypes.number,
};

const defaultProps = {
  matrixData: {},
  visibleColumns: 5,
  alwaysShowPaginationButtons: false,
  onClickCell: () => {},
  isLoading: false,
  onNextBatchRequest: null,
  nextPageText: 'NEXT SLOTS',
  shouldShowNextButton: true,
  error: '',
  hightlightMatrix: false,
  columnOffset: 0,
};

const cn = bem({ prefix: 'cmc-', block: 'pagination-matrix' });

class PaginationMatrix extends Component {
  constructor(props) {
    super(props);
    this.state = {
      visibleColumns: this.props.visibleColumns,
      columnOffset: this.props.columnOffset,
      columnsShouldIncrementBy: this.props.visibleColumns - this.props.columnOffset,
      activeSetting: {
        rowId: null,
        columnId: null,
      },
    };

    this.handleShouldPageUpdate = this.handlePageUpdate.bind(this);
    this.handleNextPageClick = this.handleNextPageClick.bind(this);
    this.handlePrevPageClick = this.handlePrevPageClick.bind(this);
    this.handleCellClick = this.handleCellClick.bind(this);
  }

  componentDidUpdate(prevProps) {
    if (!this.props.isLoading && prevProps.matrixData.headings.length !== this.props.matrixData.headings.length) {
      this.handlePageUpdate({ desiredDirection: 'next', newHeadings: prevProps.matrixData.headings });
    }
  }

  handlePrevPageClick() {
    this.handlePageUpdate({ desiredDirection: 'prev' });
  }

  handleNextPageClick() {
    this.handlePageUpdate({ desiredDirection: 'next' });
  }

  handlePageUpdate({ desiredDirection, newHeadings }) {
    const {
      onNextBatchRequest,
      matrixData: { headings },
    } = this.props;
    const { columnOffset, columnsShouldIncrementBy, visibleColumns } = this.state;

    /**
     * Based on the direction, we either add or subtract the incrementation number from the offset
     * This redraws the grid with new positional numbers
     */
    const directionalOffset = desiredDirection === 'prev' ? -columnsShouldIncrementBy : columnsShouldIncrementBy;
    const newOffset = columnOffset + directionalOffset;
    const newVisibleColumns = visibleColumns + directionalOffset;

    /**
     * After we've calculated the new positions above, we need check if there is data for those positions
     * We look to see if there is a heading in the heading array (first check in new prop heading array) for the new position
     * If there is, we redraw the table with that data
     * Otherwise we will call onNextBatchRequest to get more data
     */
    if ((!isEmpty(newHeadings) && newHeadings[newOffset]) || headings[newOffset]) {
      this.setState(() => ({
        columnOffset: newOffset,
        visibleColumns: newVisibleColumns,
      }));
    } else {
      onNextBatchRequest();
    }
  }

  handleCellClick(columnIndex, rowIndex) {
    this.props.onClickCell({ columnIndex, rowIndex });

    this.setState(() => ({
      activeSetting: {
        rowId: rowIndex,
        columnId: columnIndex,
      },
    }));
  }

  render() {
    const {
      matrixData: { headings, rows, sideHeadings },
      onNextBatchRequest,
      isLoading,
      alwaysShowPaginationButtons,
      nextPageText,
      shouldShowNextButton,
      error,
      hightlightMatrix,
    } = this.props;
    const { activeSetting, visibleColumns, columnOffset, columnsShouldIncrementBy } = this.state;
    // cell size is calculated based on how many cells should be shown
    const cellWidth = 100 / (columnsShouldIncrementBy + 1);

    const matrixHeadings = headings && (
      <MatrixHeadings
        headings={headings}
        columnOffset={columnOffset}
        visibleColumns={visibleColumns}
        activeSetting={activeSetting}
        cellWidth={cellWidth}
        hasSideHeaders={!isEmpty(sideHeadings)}
      />
    );

    const matrixRows =
      rows &&
      rows.map((row, rowIndex) => (
        <MatrixRow
          row={row}
          sideHeading={!isEmpty(sideHeadings) ? sideHeadings[rowIndex] : undefined}
          rowIndex={rowIndex}
          columnOffset={columnOffset}
          visibleColumns={visibleColumns}
          activeSetting={activeSetting}
          cellWidth={cellWidth}
          handleCellClick={this.handleCellClick}
          key={`row-${rowIndex}`}
        />
      ));

    // if headings are empty we still need to get a length.
    // We check the first row because if there are rows, the first one should be populated
    const incrementCheck = headings ? headings.length : rows && rows[0].length;
    const shouldTableDecrement = this.state.visibleColumns > columnsShouldIncrementBy;
    const shouldTableIncrement = onNextBatchRequest || incrementCheck > visibleColumns; // onNextBatchRequest !== null || incrementCheck > visibleColumns;

    const loading = (
      <div className={cn('row')}>
        <Loading height={170} fullScreen={false} />
      </div>
    );
    const matrix = (
      <Fragment>
        {error && <Notification mode='error'>{error}</Notification>}
        <div className={cn('table')}>
          {!isEmpty(headings) && matrixHeadings}
          {matrixRows}
          <div className={cn('arrow-container')}>
            {(this.props.alwaysShowPaginationButtons || shouldTableDecrement) && (
              <div className={cn('prev-arrow')}>
                <Button
                  isSecondary
                  isShort
                  buttonWidth='contentFit'
                  icon='navigate_before'
                  onClick={this.handlePrevPageClick}
                />
              </div>
            )}
            {(alwaysShowPaginationButtons || shouldTableIncrement || onNextBatchRequest) && shouldShowNextButton && (
              <div className={cn('next-arrow')}>
                <Button
                  isSecondary
                  isShort
                  buttonWidth='contentFit'
                  text={nextPageText}
                  icon='navigate_next'
                  iconMode='right'
                  onClick={this.handleNextPageClick}
                />
              </div>
            )}
          </div>
        </div>
      </Fragment>
    );
    return <div className={cn('matrix-grid', hightlightMatrix ? 'highlight' : '')}>{isLoading ? loading : matrix}</div>;
  }
}

PaginationMatrix.propTypes = propTypes;
PaginationMatrix.defaultProps = defaultProps;
export default PaginationMatrix;
