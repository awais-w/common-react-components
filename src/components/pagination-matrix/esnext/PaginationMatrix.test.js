import React from 'react';
import { shallow } from 'enzyme';
import Notification from '@cmc/cha-notification/esnext/Notification';
import PaginationMatrix from './PaginationMatrix';

const defaultProps = {
  matrixData: {
    headings: ['1', '2', '3', '4', '5'],
    sideHeadings: ['07:00', '10:00', '13:00'],
    rows: [
      [{ value: '11' }, { value: '12' }, { value: '13' }, { value: '14' }, { value: '15' }],
      [{ value: '21' }, { value: '22' }, { value: '23' }, { value: '24' }, { value: '25' }],
      [{ value: '31' }, { value: '32' }, { value: '33' }, { value: '34' }, { value: '35' }],
    ],
    columnOffset: 0,
  },
};

const setup = ({ props }) => {
  const component = shallow(<PaginationMatrix {...props} />);
  return {
    wrapper: component,
    getCornerCell: () => component.find('.cmc-pagination-matrix__blank'),
    getHeadings: () => component.find('MatrixHeadings'),
    getSideHeadings: () => component.find('.cmc-pagination-matrix__side-heading'),
    getRows: () => component.find('MatrixRow'),
    getPreviousButton: () => component.find('.cmc-pagination-matrix__prev-arrow > Button'),
    getNextButton: () => component.find('.cmc-pagination-matrix__next-arrow > Button'),
  };
};

describe('<PaginationMatrix />', () => {
  describe('with props set from default props', () => {
    const component = setup({ props: defaultProps });

    it('should match default snapshot', () => {
      expect(component.wrapper).toMatchSnapshot();
    });

    it('has the correct class', () => {
      expect(component.wrapper.hasClass('cmc-pagination-matrix__matrix-grid')).toBe(true);
    });

    describe('the rows render', () => {
      it('with the correct number of them', () => {
        expect(component.getRows().length).toBe(defaultProps.matrixData.rows.length);
      });
    });

    describe('when handleCellClick is called', () => {
      component.wrapper.instance().handleCellClick(1, 1);

      it('should pass down the active setting to the rows', () => {
        expect(component.getRows().at(0).props().activeSetting).toEqual({ rowId: 1, columnId: 1 });
      });

      it('should pass down the active setting to the headings', () => {
        expect(component.getHeadings().props().activeSetting).toEqual({ rowId: 1, columnId: 1 });
      });
    });
  });

  describe('with custom props', () => {
    describe('buttons are made to persist', () => {
      const component = setup({
        props: {
          matrixData: {
            headings: ['1'],
            rows: [['11']],
          },
          alwaysShowPaginationButtons: true,
        },
      });

      it('renders the previous button', () => {
        expect(component.getPreviousButton().length).toBe(1);
      });

      it('gives the correct props to the previous button', () => {
        const prevButtonProps = component.getPreviousButton().props();
        expect(prevButtonProps.icon).toBe('navigate_before');
        expect(prevButtonProps.isDisabled).toBe(false);
        expect(prevButtonProps.onClick).toBe(component.wrapper.instance().handlePrevPageClick);
      });

      it('renders the next button', () => {
        expect(component.getNextButton().length).toBe(1);
      });

      it('gives the correct props to the next button', () => {
        const nextButtonProps = component.getNextButton().props();
        expect(nextButtonProps.icon).toBe('navigate_next');
        expect(nextButtonProps.isDisabled).toBe(false);
        expect(nextButtonProps.onClick).toBe(component.wrapper.instance().handleNextPageClick);
      });

      it('should hide the next button if shouldShowNextButton is false', () => {
        const thisComponent = setup({ props: { ...defaultProps, shouldShowNextButton: false } });
        expect(thisComponent.getNextButton()).not.toExist();
      });
    });

    describe('when moving through the matrix by pressing the arrows', () => {
      describe('when the grid should move', () => {
        const smallMatrixProps = {
          matrixData: {
            headings: ['1', '2', '3', '4', '5'],
            sideHeadings: ['07:00', '10:00', '13:00'],
            rows: [
              [{ value: '11' }, { value: '12' }, { value: '13' }, { value: '14' }, { value: '15' }],
              [{ value: '21' }, { value: '22' }, { value: '23' }, { value: '24' }, { value: '25' }],
              [{ value: '31' }, { value: '32' }, { value: '33' }, { value: '34' }, { value: '35' }],
            ],
          },
          visibleColumns: 3,
          onNextBatchRequest: jest.fn(),
        };

        const component = setup({
          props: smallMatrixProps,
        });

        it('moves the visible columns forward when prev arrow clicked', () => {
          expect(component.getRows().at(0).props().row).toBe(smallMatrixProps.matrixData.rows[0]);
          expect(component.getRows().at(0).props().columnOffset).toBe(0);
          expect(component.getRows().at(0).props().visibleColumns).toBe(3);
          component.getNextButton().simulate('click');
          expect(component.getRows().at(0).props().row).toBe(smallMatrixProps.matrixData.rows[0]);
          expect(component.getRows().at(0).props().columnOffset).toBe(3);
          expect(component.getRows().at(0).props().visibleColumns).toBe(6);
        });

        it('moves the visible columns back when prev arrow clicked', () => {
          expect(component.getRows().at(0).props().row).toBe(smallMatrixProps.matrixData.rows[0]);
          expect(component.getRows().at(0).props().columnOffset).toBe(3);
          expect(component.getRows().at(0).props().visibleColumns).toBe(6);
          component.getPreviousButton().simulate('click');
          expect(component.getRows().at(0).props().row).toBe(smallMatrixProps.matrixData.rows[0]);
          expect(component.getRows().at(0).props().columnOffset).toBe(0);
          expect(component.getRows().at(0).props().visibleColumns).toBe(3);
        });
      });

      describe('when the grid should call props', () => {
        const component = setup({
          props: {
            matrixData: {
              headings: ['1', '2', '3', '4', '5'],
              sideHeadings: ['07:00', '10:00', '13:00'],
              rows: [
                [{ value: '11' }, { value: '12' }, { value: '13' }, { value: '14' }, { value: '15' }],
                [{ value: '21' }, { value: '22' }, { value: '23' }, { value: '24' }, { value: '25' }],
                [{ value: '31' }, { value: '32' }, { value: '33' }, { value: '34' }, { value: '35' }],
              ],
            },
            visibleColumns: 5,
            onNextBatchRequest: jest.fn(),
          },
        });

        it('should call onNextBatchRequest when there is no data', () => {
          const propSpy = jest.spyOn(component.wrapper.instance().props, 'onNextBatchRequest');
          expect(propSpy).not.toHaveBeenCalled();
          component.getNextButton().simulate('click');
          expect(propSpy).toHaveBeenCalled();
        });
      });
    });

    describe('error prop', () => {
      const component = setup({
        props: {
          error: 'This is an error message',
        },
      });
      it('shows an error message', () => {
        expect(component.wrapper.find(Notification)).toExist();
      });
    });

    describe('different props are set to empty', () => {
      describe('no headings', () => {
        const component = setup({
          props: {
            matrixData: { headings: {} },
          },
        });
        it('has no blank cell', () => {
          expect(component.getCornerCell().length).toBe(0);
        });

        it('has no header cells', () => {
          expect(component.getHeadings().length).toBe(0);
        });
      });

      describe('no side headings', () => {
        const component = setup({
          props: {
            matrixData: { sideHeadings: {} },
          },
        });
        it('has no side heading cells', () => {
          expect(component.getSideHeadings().length).toBe(0);
        });
      });

      describe('no data', () => {
        const component = setup({
          props: {
            matrixData: {},
          },
        });

        it('shouldnt render any cells', () => {
          expect(component.getHeadings().length).toBe(0);
          expect(component.getSideHeadings().length).toBe(0);
          expect(component.getRows().length).toBe(0);
        });
      });
      describe('it gets highlighted', () => {
        it('shouldnt render any cells', () => {
          const component = setup({
            props: {
              matrixData: {
                headings: ['1'],
                rows: [['11']],
              },
              hightlightMatrix: true,
            },
          });
          expect(component.wrapper.hasClass('cmc-pagination-matrix__matrix-grid--highlight')).toBe(true);
        });
      });
    });
  });
});
