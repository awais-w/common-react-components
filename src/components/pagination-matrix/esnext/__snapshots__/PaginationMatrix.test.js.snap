// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaginationMatrix /> with props set from default props should match default snapshot 1`] = `
<div
  className="cmc-pagination-matrix__matrix-grid"
>
  <div
    className="cmc-pagination-matrix__table"
  >
    <MatrixHeadings
      activeSetting={
        Object {
          "columnId": 1,
          "rowId": 1,
        }
      }
      cellWidth={16.666666666666668}
      columnOffset={0}
      hasSideHeaders={true}
      headings={
        Array [
          "1",
          "2",
          "3",
          "4",
          "5",
        ]
      }
      visibleColumns={5}
    />
    <MatrixRow
      activeSetting={
        Object {
          "columnId": 1,
          "rowId": 1,
        }
      }
      cellWidth={16.666666666666668}
      columnOffset={0}
      handleCellClick={[Function]}
      key="row-0"
      row={
        Array [
          Object {
            "value": "11",
          },
          Object {
            "value": "12",
          },
          Object {
            "value": "13",
          },
          Object {
            "value": "14",
          },
          Object {
            "value": "15",
          },
        ]
      }
      rowIndex={0}
      sideHeading="07:00"
      visibleColumns={5}
    />
    <MatrixRow
      activeSetting={
        Object {
          "columnId": 1,
          "rowId": 1,
        }
      }
      cellWidth={16.666666666666668}
      columnOffset={0}
      handleCellClick={[Function]}
      key="row-1"
      row={
        Array [
          Object {
            "value": "21",
          },
          Object {
            "value": "22",
          },
          Object {
            "value": "23",
          },
          Object {
            "value": "24",
          },
          Object {
            "value": "25",
          },
        ]
      }
      rowIndex={1}
      sideHeading="10:00"
      visibleColumns={5}
    />
    <MatrixRow
      activeSetting={
        Object {
          "columnId": 1,
          "rowId": 1,
        }
      }
      cellWidth={16.666666666666668}
      columnOffset={0}
      handleCellClick={[Function]}
      key="row-2"
      row={
        Array [
          Object {
            "value": "31",
          },
          Object {
            "value": "32",
          },
          Object {
            "value": "33",
          },
          Object {
            "value": "34",
          },
          Object {
            "value": "35",
          },
        ]
      }
      rowIndex={2}
      sideHeading="13:00"
      visibleColumns={5}
    />
    <div
      className="cmc-pagination-matrix__arrow-container"
    />
  </div>
</div>
`;
